docker/Dockerfile

```diff
@@ -13,47 +13,140 @@
+RUN apt-get update \
+    && apt-get install -y software-properties-common \
+    && apt-get update && apt-get install -y --no-install-recommends \
+    autoconf \
+    bison \
+    build-essential \
+    ca-certificates \
+    ccache \
+    clang-format-7 \
+    clang-tidy \
+    cron \
+    curl \
+    dnsutils \
+    gawk \
+    gdb \
+    git \
+    htop \
+    iproute2 \
+    lcov \
+    libcurl4-openssl-dev \
+    libsecp256k1-dev \
+    libssl-dev \
+    libtool \
+    libxml2-utils \
+    logrotate \
+    net-tools \
+    ninja-build \
+    nload \
+    ocl-icd-opencl-dev \
+    openssh-client \
+    pkg-config \
+    python3-dev \
+    python3-pip \
+    rsync \
+    rsyslog \
+    tar \
+    trickle \
+    unzip \
+    vim \
+    wget \
+    zip \
+    && apt-get remove -y cmake python2  && apt-get autoremove -y
+
+FROM deps AS builder
+
+# Cargo
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs -o install_script.sh && \
+    sh install_script.sh -y
+ENV PATH="/root/.cargo/bin:${PATH}"
+
+RUN cargo install sccache
+ENV SCCACHE_CACHE_SIZE="1G"
+ENV SCCACHE_DIR="/sccache"
+ENV RUSTC_WRAPPER="/root/.cargo/bin/sccache"
+
+# CMake 3.24
+ARG CMAKE_VERSION=3.24.2
+RUN wget https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/cmake-${CMAKE_VERSION}-Linux-x86_64.sh \
+    && echo "739d372726cb23129d57a539ce1432453448816e345e1545f6127296926b6754 cmake-${CMAKE_VERSION}-Linux-x86_64.sh" | sha256sum -c \ 
+    && mkdir -p "${HOME}"/.local \
+    && bash ./cmake-${CMAKE_VERSION}-Linux-x86_64.sh --skip-license --prefix="${HOME}"/.local/ \
+    && "${HOME}"/.local/bin/cmake --version \
+    && rm cmake-${CMAKE_VERSION}-Linux-x86_64.sh
+ENV PATH="/root/.local/bin:${PATH}"
+
+# Setup ccache
+RUN ln -s "$(which ccache)" /usr/local/bin/gcc \
+  && ln -s "$(which ccache)" /usr/local/bin/g++ \
+  && ln -s "$(which ccache)" /usr/local/bin/cc \
+  && ln -s "$(which ccache)" /usr/local/bin/c++
+
+# This tag must be equivalent to the hash specified by "builtin-baseline" in vcpkg.json
+ARG VCPKG_COMMIT_OR_TAG=2022.09.27
+ARG VCPKG_ROOT=/vcpkg
 
-# run a copy -L to unfold the symlinkes, and strip all exes
-RUN mkdir -p /scilla/0/bin2/ && cp -L /scilla/0/bin/* /scilla/0/bin2/ && strip /scilla/0/bin2/*
+RUN ccache -p && ccache -z
 
-FROM zilliqa/zilliqa-ccache:${ZILLIQA_CCACHE_VERSION} AS builder
+# If COMMIT_OR_TAG is a branch name or a tag, clone a shallow copy which is
+# faster; if this fails, just clone the full repo and checkout the commit.
+RUN git clone https://github.com/microsoft/vcpkg ${VCPKG_ROOT} \
+  && git -C ${VCPKG_ROOT} checkout ${VCPKG_COMMIT_OR_TAG} \
+  && ${VCPKG_ROOT}/bootstrap-vcpkg.sh
 
-# Manually input tag or commit, can be overwritten by docker build-args
-ARG COMMIT_OR_TAG=v8.4.0
-ARG REPO=https://github.com/Zilliqa/Zilliqa.git
-ARG SOURCE_DIR=/root/zilliqa
-ARG BUILD_DIR=/root/zilliqa/build
-ARG INSTALL_DIR=/usr/local
-ARG VCPKG_ROOT=/vcpkg
+WORKDIR /zilliqa
```

Comment:
As mentioned, not great to put this under /

Reply:
I tested quite a few times and it seemed to work fine. What was the issue you were thinking of?

Reply:
The ones I explained previously. Setting / as the basedir can have issues matching the correct file in the cache. In addition, it's explicitly stated in the ccache documentation:

_Don’t use / as the base directory since that will make ccache also rewrite paths to system header files, which typically is contraproductive_.

When you said you tried and it worked, did you try this with a warm cache? If so, what cache hit rate did you get?

Reply:
The base dir here is `/zilliqa` though so surely that caveat doesn't apply?

Here's the logs from a run with a warm cache after I change a source file:
```
#32 0.349 Statistics zeroed
#32 0.349 ccache status
#32 0.350 cache directory                     /ccache
#32 0.350 primary config                      /ccache/ccache.conf
#32 0.350 secondary config      (readonly)    /etc/ccache.conf
#32 0.350 stats updated                       Thu Jan  5 15:37:35 2023
#32 0.350 stats zeroed                        Thu Jan  5 15:37:35 2023
#32 0.350 cache hit (direct)                     0
#32 0.350 cache hit (preprocessed)               0
#32 0.350 cache miss                             0
#32 0.350 cache hit rate                      0.00 %
#32 0.350 cleanups performed                     0
#32 0.350 files in cache                       957
#32 0.350 cache size                           4.1 GB
#32 0.350 max cache size                       5.0 GB
...
#32 37.16 ccache status
#32 37.17 cache directory                     /ccache
#32 37.17 primary config                      /ccache/ccache.conf
#32 37.17 secondary config      (readonly)    /etc/ccache.conf
#32 37.17 stats updated                       Thu Jan  5 15:38:03 2023
#32 37.17 stats zeroed                        Thu Jan  5 15:37:35 2023
#32 37.17 cache hit (direct)                   169
#32 37.17 cache hit (preprocessed)              13
#32 37.17 cache miss                             0
#32 37.17 cache hit rate                    100.00 %
#32 37.17 called for link                       39
#32 37.17 called for preprocessing              13
#32 37.17 no input file                          6
#32 37.17 cleanups performed                     0
#32 37.17 files in cache                       957
#32 37.17 cache size                           4.1 GB
#32 37.17 max cache size                       5.0 GB
```

Reply:
Why would the Jenkins pipeline build from a different location? The Dockerfile is self-contained, so we'll only ever build inside `/zilliqa`. I think it would work from a separate location since the ccache is shared, but I don't see why that would ever matter.

Reply:
Because the CI Jenkinsfile job works differently. Always has. You can see the Jenkinsfile for details and also some of the logs to get an idea (e.g. https://ci.pipeline.z7a.xyz/blue/rest/organizations/jenkins/pipelines/ZilliqaCIJenkinsfile/branches/valeryz-rc-3/runs/2/log/?start=0). It essentially builds by invoking the build script inside a container and doesn't simply builds a docker image the same way that Zilliqa Private Image Manual Trigger does.

Of course ccache is shared, but previously it used to have issues matching the same file if it was from a different location. I suggest you give it a try. If it works - great. If not, will need to be fixed.

Reply:
Hmm, looking at the `Jenkinsfile` and those logs, I don't think we're using docker at all for that CI pipeline. We invoke `./ci_build.sh` inside a container, but that image is `zilliqa/zilliqa-ccache:v0.0.0`.

Frankly, I think we should throw away our CI soon and rebuild it so it isn't intractable.

The other Jenkins pipelines seem to just build an image with `docker build` (via `make`) so I think they should work fine?

Reply:
That's precisely what I've been saying. You can read again one of my first comments here (https://github.com/Zilliqa/Zilliqa/pull/3299#pullrequestreview-1221685848) which was precisely about this and why I asked whether you'd tested both scenarios.

Both ci_build.sh & the Dockerfile (through docker build) eventually invoke build.sh but this will occur from different locations. The whole point of the ccache image was to provide a warm(er) ccache to speed up these jobs as well. So the question of whether this works as expected after your changes.

I don't like how we have these different ways of building things (Jenkinsfile vs. Dockerfile) any more than you do but it's a given. We can't break things or make them worse. 
@mauromedda has been working at improving this so indeed we should simplify this going forward. Another idea for improvement would be to provide a shared ccache to all builds. This however requires more work and coordination with the DevOps team.

Anyway, it feels a bit like we're going in circles here.

ccache currently doesn't provide a huge benefit for the Jenkinsfile builds (as there's no shared cache) but it does provide some and beside being the ground work for future improvements we really want to reduce build times which are still pretty high. Therefore, no reason to deprecate this.

Reply:
The Jenkins job works - https://ci.pipeline.z7a.xyz/blue/organizations/jenkins/ZilliqaCIJenkinsfile/detail/PR-3299/4/pipeline/61

This PR didn't chance any of the build scripts, so it shouldn't change anything outside of Docker builds.

Reply:
Ah sorry, I was missing the point, thank you :laughing: 

So I guess the open question is what to use as the base image for the native Jenkins build? It currently uses `zilliqa-ccache`, so we should either keep the Dockerfile for that around, or change the base image to something else?

With a static base image containing the cache, I'd assume that caching performance will go down with time, because the cache will get more and more stale, without being updated. Ideally, the cache directory would be mounted as a persistent volume, so that it gets updated by each subsequent build. cc @mauromedda ?

Is this something better to fix in a subsequent PR though? This shouldn't break anything, since the existing image will still exist in Dockerhub?